### PR TITLE
Preparations to make RIOT 64-bit ready

### PR DESCRIPTION
**Pull request by OlegHahm**
Adopted from #19890

### Contribution description

Following @benpicco's proposal I handpicked the patches from #19890 that only prepare the code base for 64bit support.

All changes are done by @fzi-haxel.

### Testing procedure

All tests and applications should remain unchanged for existing platforms.

### CODE REVIEW

File: pkg/flashdb/patches/0002-64bit-fix.patch [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+diff --git a/src/fdb_kvdb.c b/src/fdb_kvdb.c
+index 393c463..7905fda 100644
+--- a/src/fdb_kvdb.c
++++ b/src/fdb_kvdb.c
+@@ -341,7 +341,7 @@ static fdb_err_t read_kv(fdb_kvdb_t db, fdb_kv_t kv)
+     kv->status = (fdb_kv_status_t) _fdb_get_status(kv_hdr.status_table, FDB_KV_STATUS_NUM);
+     kv->len = kv_hdr.len;
+ 
+-    if (kv->len == ~0UL || kv->len > db_max_size(db) || kv->len < KV_HDR_DATA_SIZE) {
++    if (kv->len == UINT32_MAX || kv->len > db_max_size(db) || kv->len < KV_HDR_DATA_SIZE) {
```
Comment (benpicco) [inline pkg/flashdb/patches/0002-64bit-fix.patch:19]: I think you can drop this when #20155 is merged

File: pkg/micro-ecc/patches/0001-No-128-Bit-types-on-gcc.patch [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+diff --git a/types.h b/types.h
+index 7cb1a28..0a0ff2d 100644
+--- a/types.h
++++ b/types.h
+@@ -51,7 +51,7 @@
+     #define uECC_WORD_SIZE 4
+ #endif
+ 
+-#if defined(__SIZEOF_INT128__) || ((__clang_major__ * 100 + __clang_minor__) >= 302)
++#if ((__clang_major__ * 100 + __clang_minor__) >= 302)
```
Comment (benpicco) [inline pkg/micro-ecc/patches/0001-No-128-Bit-types-on-gcc.patch:19]: that one I don't understand

File: pkg/micropython/patches/0001-Ensure-mp-int-and-uint-have-pointer-size.patch [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ #include "irq.h"
+ 
+@@ -96,8 +97,8 @@
+ 
+ #define MICROPY_MODULE_FROZEN_STR       (1)
+ 
+-typedef int mp_int_t;       // must be pointer size
+-typedef unsigned mp_uint_t; // must be pointer size
++typedef intptr_t mp_int_t;       // must be pointer size
++typedef uintptr_t mp_uint_t; // must be pointer size
```
Comment (benpicco) [inline pkg/micropython/patches/0001-Ensure-mp-int-and-uint-have-pointer-size.patch:29]: Since we are using [our own fork](https://github.com/kaspar030/micropython) anyway, you might as well open the PR there.

File: pkg/tinydtls/patches/0002-fix-printf-warning-for-64-bit-systems.patch [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+--- a/dtls_debug.h
++++ b/dtls_debug.h
+@@ -139,8 +139,8 @@ void dtls_dsrv_log_addr(log_t level, const char *name, const session_t *addr);
+ #define dtls_notice(...) LOG_INFO(__VA_ARGS__)
+ #define dtls_info(...) LOG_INFO(__VA_ARGS__)
+ #define dtls_debug(...) LOG_DEBUG(__VA_ARGS__)
+-#define dtls_debug_hexdump(name, buf, length) { if (LOG_DEBUG <= LOG_LEVEL) { LOG_DEBUG("-= %s (%zu bytes) =-\n", name, length); od_hex_dump(buf, length, 0); }}
+-#define dtls_debug_dump(name, buf, length) { if (LOG_DEBUG <= LOG_LEVEL) { LOG_DEBUG("%s (%zu bytes):", name, length); od_hex_dump(buf, length, 0); }}
++#define dtls_debug_hexdump(name, buf, length) { if (LOG_DEBUG <= LOG_LEVEL) { LOG_DEBUG("-= %s (%zu bytes) =-\n", name, (size_t) length); od_hex_dump(buf, length, 0); }}
++#define dtls_debug_dump(name, buf, length) { if (LOG_DEBUG <= LOG_LEVEL) { LOG_DEBUG("%s (%zu bytes):", name, (size_t) length); od_hex_dump(buf, length, 0); }}
```
Comment (benpicco) [inline pkg/tinydtls/patches/0002-fix-printf-warning-for-64-bit-systems.patch:21]: We usually get our patches merged upstream for this quite well, so please try to open a PR there. (Reminds me that I can delete `0001-dtls_debug-mock-dsrv_log-when-NDEBUG.patch` now since that one [got merged](https://github.com/eclipse/tinydtls/pull/189))

**Review by benpicco — COMMENTED** [review]
This all looks reassuringly straightforward!

File: pkg/micro-ecc/patches/0001-No-128-Bit-types-on-gcc.patch (hunk 2 above)
Comment (OlegHahm) [reply pkg/micro-ecc/patches/0001-No-128-Bit-types-on-gcc.patch:19]: good question - @fzi-haxel?

File: pkg/micropython/patches/0001-Ensure-mp-int-and-uint-have-pointer-size.patch (hunk 3 above)
Comment (OlegHahm) [reply pkg/micropython/patches/0001-Ensure-mp-int-and-uint-have-pointer-size.patch:29]: True - but actually unrelated to this PR, isn't it?

File: pkg/micro-ecc/patches/0001-No-128-Bit-types-on-gcc.patch (hunk 2 above)
Comment (fzi-haxel) [reply pkg/micro-ecc/patches/0001-No-128-Bit-types-on-gcc.patch:19]: On x86_64, `__SIZEOF_INT128__` is defined, therefore `__int128` is used, but this leads to this GCC warning:
```
build/pkg/micro-ecc/types.h:98:18: error: ISO C does not support ‘__int128’ types [-Werror=pedantic]
   98 | typedef unsigned __int128 uECC_dword_t;
```

In retrospect, simply not using `__int128` does not seem to be the most elegant solution. Another option would be:
```
__extension__ typedef unsigned __int128 uECC_dword_t;
```

However, the GCC documentation points out the following when using `__extension__`
> only system header files should use these escape routes.

What do you think?

Comment (OlegHahm) [reply pkg/micro-ecc/patches/0001-No-128-Bit-types-on-gcc.patch:19]: I'm a bit confused: clang accepts `__int128` even though it is not part of the C standard but gcc does not?

What I also do not understand: if `__int128` is a built-in data type on gcc (>= 4.1) wouldn't I still get the warning even with the typedef? (Which definitiv would be actually used?)

**Comment by OlegHahm** [comment]
Btw @fzi-haxel by no means I wanted to take over your contribution. I just got excited about your PR and the idea to have (finally) 64 bit support (for native) in RIOT and I thought that @benpicco had a good point that we could potentially speed up things by splitting the PR. Since the PR did seem stall somewhat, I just moved forward without asking you first. I hope that was okay with you!?

File: pkg/micropython/patches/0001-Ensure-mp-int-and-uint-have-pointer-size.patch (hunk 3 above)
Comment (OlegHahm) [reply pkg/micropython/patches/0001-Ensure-mp-int-and-uint-have-pointer-size.patch:29]: Made a PR in the fork.

File: pkg/tinydtls/patches/0002-fix-printf-warning-for-64-bit-systems.patch (hunk 4 above)
Comment (OlegHahm) [reply pkg/tinydtls/patches/0002-fix-printf-warning-for-64-bit-systems.patch:21]: PR created.

File: pkg/micro-ecc/patches/0001-No-128-Bit-types-on-gcc.patch (hunk 2 above)
Comment (fzi-haxel) [reply pkg/micro-ecc/patches/0001-No-128-Bit-types-on-gcc.patch:19]: If `-Wpedantic` is enabled, yes.
Example: https://godbolt.org/z/bsYq49jGY

Comment (OlegHahm) [reply pkg/micro-ecc/patches/0001-No-128-Bit-types-on-gcc.patch:19]: So this typedef wouldn't help to get rid of the warning anyway.

**Comment by fzi-haxel** [comment]
> Btw @fzi-haxel by no means I wanted to take over your contribution. I just got excited about your PR and the idea to have (finally) 64 bit support (for native) in RIOT and I thought that @benpicco had a good point that we could potentially speed up things by splitting the PR. Since the PR did seem stall somewhat, I just moved forward without asking you first. I hope that was okay with you!?

No, it's great, I was going to do it anyway, but I was busy for a while and then it kind of fell through the cracks...
And I really appreciate the support.

File: pkg/micro-ecc/patches/0001-No-128-Bit-types-on-gcc.patch (hunk 2 above)
Comment (fzi-haxel) [reply pkg/micro-ecc/patches/0001-No-128-Bit-types-on-gcc.patch:19]: Using `__extension__` would get rid of the warning.

Comment (fzi-haxel) [reply pkg/micro-ecc/patches/0001-No-128-Bit-types-on-gcc.patch:19]: Are you sure? I just tested it with (13.2.0) and didn't get this warning.

But GCC 13 now gives this warning during linking
```
/usr/bin/ld: warning: /work/tests/pkg/micro-ecc/bin/native/cpu/tramp.o: missing .note.GNU-stack section implies executable stack
```

Comment (OlegHahm) [reply pkg/micro-ecc/patches/0001-No-128-Bit-types-on-gcc.patch:19]: Ah, no, sorry, I think there was a PEBCAK issue (I always confuse the order of arguments for typedef).

File: drivers/atwinc15x0/atwinc15x0_netdev.c [hunk 5]
```diff
@@ -646,7 +646,7 @@ static int _atwinc15x0_get(netdev_t *netdev, netopt_t opt, void *val,
     assert(dev == atwinc15x0);
 
     DEBUG("%s dev=%p opt=%u val=%p max_len=%u\n", __func__,
-          (void *)netdev, opt, val, max_len);
+          (void *)netdev, opt, val, (unsigned)max_len);
```
Comment (kfessel) [inline drivers/atwinc15x0/atwinc15x0_netdev.c:649]: is `sizeof(unsigned) >= sizeof(size_t)` --  i am not sure (ul seem like the least maybe sizeof(void*)-> PRI*PTR )
even thought one would assume any thing we actually got there should be fit for mcu there

File: tests/sys/events/main.c [hunk 6]
```diff
@@ -62,7 +62,7 @@ static void callback(event_t *arg)
     order++;
     expect(order == 4);
     expect(arg == &event);
-    printf("triggered 0x%08x\n", (unsigned)arg);
+    printf("triggered 0x%08" PRIuPTR "\n", (uintptr_t)arg);
```
Comment (kfessel) [inline tests/sys/events/main.c:65]: why do you change from hexadecialmal to decimal printing ( with 0x in front)?

File: drivers/atwinc15x0/atwinc15x0_netdev.c (hunk 5 above)
Comment (OlegHahm) [reply drivers/atwinc15x0/atwinc15x0_netdev.c:649]: Sorry, I don't understand your comment. Can you rephrase?

**Comment by kfessel** [comment]
some of the (debug)prints are strange might compile but not show what I would expect

File: drivers/atwinc15x0/atwinc15x0_netdev.c (hunk 5 above)
Comment (kfessel) [reply drivers/atwinc15x0/atwinc15x0_netdev.c:649]: There is a max_len in size_t going in. With x86_64  this is of size 64Bit but this cast is to 32Bit - only some of the bits are used in that print / DEBUG -> this might hide the exact issue one is debugging.

PRI*PTR might be the right size print in most cases ( in some it might be to large) or a little less wordy lu or ul  which would might be fit for every thing we do (not sure).

lu is also the implementation the newlib defaults to for printing so no code size increase of any arch using that.

For u I know of one case it is to small to catch all of size_t.

The second part of the sentence is about the assumption of fzi-haxel that max_len is unlikely to exceed u on x86_64 because most of our systems are smaller and therefore any valid len is likely to fit into unsigned

I picked on this by on example there are multiple instances of this change some of them are more likely behave wrong than in this static function.

File: tests/sys/events/main.c [hunk 7]
```diff
@@ -176,12 +177,12 @@ int main(void)
...
      * initialized queue */
     event_queue_t queue = EVENT_QUEUE_INIT;
-    printf("posting 0x%08x\n", (unsigned)&event);
+    printf("posting 0x%08" PRIuPTR "\n", (uintptr_t)&event);
     event_post(&queue, &event);
 
-    printf("posting 0x%08x\n", (unsigned)&event2);
+    printf("posting 0x%08" PRIuPTR "\n", (uintptr_t)&event2);
     event_post(&queue, &event2);
-    printf("canceling 0x%08x\n", (unsigned)&event2);
```
Comment (kfessel) [inline tests/sys/events/main.c:184]: u for x

File: tests/sys/events/main.c [hunk 8]
```diff
@@ -94,7 +94,8 @@ static void timed_callback(void *arg)
     uint32_t now = xtimer_now_usec();
 #endif
     expect((now - before >= 100000LU));
-    printf("triggered timed callback with arg 0x%08x after %" PRIu32 "us\n", (unsigned)arg, now - before);
```
Comment (kfessel) [inline tests/sys/events/main.c:97]: ufor x

File: tests/sys/events/main.c (hunk 7 above)
Comment (fzi-haxel) [reply tests/sys/events/main.c:184]: Also a typo

File: tests/net/gnrc_sixlowpan_frag_minfwd/main.c [hunk 9]
```diff
@@ -624,10 +624,16 @@ static void test_minfwd_forward__ENOMEM__netif_hdr_build_fail(void)
     gnrc_pktsnip_t *pkt, *frag, *filled_space;
 
     vrbe->super.arrival = xtimer_now_usec();
+
+    /* offset: 2 * sizeof(gnrc_pktsnip_t) + movement due to mark */
+#if (__SIZEOF_POINTER__ == 8)
+        size_t offset = 155U;
+#else
+        size_t offset = 115U;
+#endif
```
[the remark this answers is not in the document]
Comment (kfessel) [reply tests/net/gnrc_sixlowpan_frag_minfwd/main.c:633]: could you point me to a code / documentation line where i could see what is meant by 'mark'

[the remark this answers is not in the document]
Comment (fzi-haxel) [reply tests/net/gnrc_sixlowpan_frag_minfwd/main.c:633]: This was not my comment, I just moved it, but my understanding is that it essentially refers to this [`sys/include/net/gnrc/pktbuf.h:130`.](https://github.com/RIOT-OS/RIOT/blob/b0261f31abc35d9e33405abccd0a23e8be56494d/sys/include/net/gnrc/pktbuf.h#L97)

File: pkg/micro-ecc/patches/0001-No-128-Bit-types-on-gcc.patch (hunk 2 above)
Comment (fzi-haxel) [reply pkg/micro-ecc/patches/0001-No-128-Bit-types-on-gcc.patch:19]: Opened an issue upstream with some suggested solutions

**Comment by fzi-haxel** [comment]
I pushed my fixups as a PR to Oleg's branch.
It also includes a small bug fix for a recent change in unittests. If I should do a separate PR, just say so.

File: tests/net/gnrc_sixlowpan_frag_minfwd/main.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (kfessel) [reply tests/net/gnrc_sixlowpan_frag_minfwd/main.c:633]: Thank you for providing the hint.

sadly i am not sure about this change gnrc_pktbuf_add is usually used on network pakets/buffers  that should have a similar structure on 8 to 64Bit (since they should read /write the same).

**Comment by OlegHahm** [comment]
> I pushed my fixups as a PR to Oleg's branch. It also includes a small bug fix for a recent change in unittests. If I should do a separate PR, just say so.

Merged

File: tests/net/gnrc_sixlowpan_frag_minfwd/main.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (OlegHahm) [reply tests/net/gnrc_sixlowpan_frag_minfwd/main.c:633]: Maybe @miri64 can shed some light on this?

[the remark this answers is not in the document]
Comment (fzi-haxel) [reply tests/net/gnrc_sixlowpan_frag_minfwd/main.c:633]: The network buffers are still the same, but they are combined with `gnrc_pktsnip_t` and `_unused_t` elements in the "static pktbuf heap" (`_static_buf[CONFIG_GNRC_PKTBUF_SIZE]`) and all elements are aligned to the size of `_unused_t`, making it architecture dependent.
The test tries to trigger a `ENOMEM` error in `gnrc_sixlowpan_frag_minfwd_forward` due to a full buffer.
I did some testing and the state of the buffer after the two adds and the mark operation should be
```
S1|D1|S2|U|S3|D2|D3
```
with 
```
Sx: gnrc_pktsnip_t - size: _aligned(sizeof(gnrc_pktsnip_t))
Dx: data buffer - size: _aligned(bufferSize)
U: Free space that is too small for the `gnrc_sixlowpan_frag_minfwd_forward` call
```

After some calculations, the offset should be
```
size_t first_pkt_snip = _align(sizeof(gnrc_pktsnip_t));
size_t test_pkt_size = _align(sizeof(gnrc_pktsnip_t)) + _align(sizeof(_test_nth_frag));
size_t marked_pkt_size = _align(sizeof(gnrc_pktsnip_t)) + _align(sizeof(sixlowpan_frag_n_t)) +  
     _align(sizeof(_test_nth_frag) - sizeof(sixlowpan_frag_n_t));
size_t offset = first_pkt_snip + test_pkt_size + marked_pkt_size;

TEST_ASSERT_NOT_NULL((filled_space = gnrc_pktbuf_add(
        NULL, NULL,
        CONFIG_GNRC_PKTBUF_SIZE  - offset,
        GNRC_NETTYPE_UNDEF
    )));
```
However, one problem with this code is that we can't call `_align' from the test.

File: tests/unittests/tests-hashes/tests-hashes-sha512.c [hunk 10]
```diff
@@ -348,12 +348,12 @@ static void test_hashes_sha512_hash_update_twice(void)
...
 
     sha512_init(&sha512);
-    sha512_update(&sha512, (uint8_t*)teststring, sizeof(teststring));
+    sha512_update(&sha512, (uint8_t*)teststring, strlen(teststring));
     sha512_final(&sha512, hash_update_once);
 
     sha512_init(&sha512);
     sha512_update(&sha512, (uint8_t*)teststring, 3);
-    sha512_update(&sha512, (uint8_t*)&teststring[3], sizeof(teststring)-3);
+    sha512_update(&sha512, (uint8_t*)&teststring[3], strlen(teststring)-3);
```
Comment (kfessel) [inline tests/unittests/tests-hashes/tests-hashes-sha512.c:356]: since this is not related to the 64 Bit preparation -> please make this a separate PR  this should go through realy fast -- ping me if no one picks that up within a day 

i also would prefer this:
```suggestion
    sha512_update(&sha512, (uint8_t*)&teststring[3], strlen(&teststring[3]));
```

**Comment by kfessel** [comment]
> I pushed my fixups as a PR to Oleg's branch. It also includes a small bug fix for a recent change in unittests. If I should do a separate PR, just say so.

please do

**Comment by OlegHahm** [comment]
I lost a little bit track of what is still open here to be addressed.

**Comment by fzi-haxel** [comment]
> I lost a little bit track of what is still open here to be addressed.

Done:
- `pkg/flashdb/patches/0002-64bit-fix.patch`
   - [PR1](https://github.com/armink/FlashDB/pull/243), [PR2](https://github.com/armink/FlashDB/pull/267), [PR3](https://github.com/armink/FlashDB/pull/268) merged upstream and version already in bumped RIOT ([PR](https://github.com/RIOT-OS/RIOT/pull/20155)) -> Removed patch file from this PR
- `tests/periph/flashpage/main.c`, `tests/net/ieee802154_submac/main.c`, `drivers/lcd/lcd.c` & `drivers/atwinc15x0/atwinc15x0_netdev.c`
   - Print `size_t` variable: changed cast from `unsigned` to `unsigned long`
- `tests/sys/events/main.c` & `core/mbox.c`
   - Typo: `PRIuPTR` instead of `PRIxPTR`
- `tests/unittests/tests-hashes/tests-hashes-sha512.c`
   - fixed in separate [PR](https://github.com/RIOT-OS/RIOT/pull/20165) -> removed from this PR

In progress:
- `pkg/micro-ecc/patches/0001-No-128-Bit-types-on-gcc.patch`: Open [issue](https://github.com/kmackay/micro-ecc/issues/222)
   - Currently the patch just disables `__int128` for GCC
   - No response yet (last activity in April)
   - Might take a while
- `pkg/micropython/patches/0001-Ensure-mp-int-and-uint-have-pointer-size.patch`: Open [PR](https://github.com/kaspar030/micropython/pull/4) 
   - No response yet (last activity 4 years ago)
   - We should talk about maintaining this repository in general. It's 4 years behind upstream.
- `pkg/tinydtls/patches/0002-fix-printf-warning-for-64-bit-systems.patch`: Open [PR](https://github.com/eclipse/tinydtls/pull/217)
   - Should be merge in the foreseeable future

Open points:
- Use `unsigned long` in print statements for `size_t` throughout the RIOT code base. 
  - Requires a lot of code review and changes to the `CODING_CONVENTIONS.md`
  - Impact on current architectures:
    - 8-bit: `sizeof(size_t) == sizeof(unsigned) == 16`, `sizeof(unsigned long) == 32` -> Small increase in stack usage and runtime
    - 32-bit: `sizeof(size_t) == sizeof(unsigned) == sizeof(unsigned long) == 32` -> No impact
    - x86_64: `sizeof(size_t) == sizeof(unsigned long) == 64`, `sizeof(unsigned) == 32` -> Avoids printing the wrong value, if the value is above 4294967295
    - Furthermore, in most cases we can see that the value of the variable will not exceed 32 bits (basically all packet sizes, uses of ARRAY_SIZE, etc.).
  - I suggest creating a separate issue or PR to discuss the impact on 8-bit architectures and get some support for the code review.
 
- `tests/net/gnrc_sixlowpan_frag_minfwd/main.c`: 
   - Architecture specific magic numbers (`115U` and `155U`) to fill the buffer in `test_minfwd_forward__ENOMEM__netif_hdr_build_fail`.
   - I could copy the aligment code to the test and use the calculation I posted here. But it is not a very elegant solution.
   - Perhaps @miri64 has a better suggestion to improve this.
- Rebase to master due to merge conflict
   - This line in `tests/periph/timer/main.c` is the problem:
```
-    if (timer_init(TIMER_DEV(num), TIMER_SPEED, cb, (void *)(uintptr_t)(COOKIE * num)) < 0) {
+    if (timer_init(TIMER_DEV(num), timer_freq, cb, (void *)(uintptr_t)(COOKIE * num)) != 0) {
```

I would appreciate some feedback on the ongoing and open issues. 
Finally, can I use `git rebase -i --autosquash` (or alternatively just a `git rebase`) to fix the merge problem?

**Comment by OlegHahm** [comment]
Thanks for the summary:
> Use unsigned long in print statements for size_t throughout the RIOT code base.
> * I suggest creating a separate issue or PR to discuss the impact on 8-bit architectures and get some support for the code review.

Sounds valid to me.

Regarding the magic numbers in `tests/net/gnrc_sixlowpan_frag_minfwd/main.c`:
I agree that these numbers should be replaced with something more meaningful but that is basically unrelated to this PR.

Should I do the rebase?

**Comment by miri64** [comment]
> Regarding the magic numbers in `tests/net/gnrc_sixlowpan_frag_minfwd/main.c`:
> I agree that these numbers should be replaced with something more meaningful but that is basically unrelated to this PR.

Not sure, why I wrote in the comment where the value came from and on what type the magic number depends on, but then did not use `sizeof(gnrc_pktsnip_t)` in the code...

**Comment by fzi-haxel** [comment]
> Should I do the rebase?

Sure

**Comment by OlegHahm** [comment]
> > Should I do the rebase?
> 
> Sure

Done

**Comment by OlegHahm** [comment]
Is there actually a reason against introducing macros for printing `size_t` variables? Like `PRISIZE`.

**Comment by fzi-haxel** [comment]
> Is there actually a reason against introducing macros for printing `size_t` variables? Like `PRISIZE`.

I don't think so, but it still doesn't save us from having to go through the entire code base :)
We could do something like [autoconf](https://www.gnu.org/software/autoconf-archive/ax_printf_size_t.html) can do. But I'm not sure what kind of header it should be put in.

```
#include <stdint.h>
#include <stddef.h>
#include <limits.h>

#if UINT_MAX == SIZE_MAX
#define PRI_SIZE_T_MODIFIER ""
#elif ULONG_MAX == SIZE_MAX
#define PRI_SIZE_T_MODIFIER "l"
#else
#error Unsupported size_t length
#endif

#define PRIdSIZE PRI_SIZE_T_MODIFIER "d"
#define PRIiSIZE PRI_SIZE_T_MODIFIER "i"
#define PRIoSIZE PRI_SIZE_T_MODIFIER "o"
#define PRIuSIZE PRI_SIZE_T_MODIFIER "u"
#define PRIxSIZE PRI_SIZE_T_MODIFIER "x"
#define PRIXSIZE PRI_SIZE_T_MODIFIER "X"
```

**Comment by fzi-haxel** [comment]
I added an RFC for the `size_t` format issue and added two fixes in a [PR](https://github.com/OlegHahm/RIOT/pull/23) to Oleg's branch.

- Fix for failed murdock test 
  Basically, this is an alternative workaround for the "assert is always true" warning when `FLASH_BASE == 0`. Upstream casts the address to `int`, which I first changed to a `uintptr_t' cast, but this retriggered the warning.

- Fix for clang lwip warning
  GCC didn't complain about this, and basically by sheer coincidence it didn't crash the 64-bit tests. But compiling with LLVM threw a warning. The patch changes `int res` to `ssize_t res` and still casts the signed `res` to an unsigned `size_t` to report the bytes written in the lwip function call, but since the `buf->p->len` is of type `unsigned short`, this should still not cause problems.

Feel free to give feedback to the RFC and review the changes.

**Comment by OlegHahm** [comment]
I rebased on current master (including @fzi-haxel's `size_t` formatters) and fixed the resulting merge conflicts.

File: pkg/micropython/patches/0001-Ensure-mp-int-and-uint-have-pointer-size.patch (hunk 3 above)
Comment (OlegHahm) [reply pkg/micropython/patches/0001-Ensure-mp-int-and-uint-have-pointer-size.patch:29]: PR is merged, waiting for #20208 to get merged before we can drop this patch.

**Comment by fzi-haxel** [comment]
I pushed two more PRs to Olegs branch:

1. [examples/dtls-sock: Revert changes due to separate PR](https://github.com/OlegHahm/RIOT/pull/24)
    Created #20196 as this is technically a general issue. It would be very helpful if someone with a compatible board could test it.
    (This should hopefully fix the murdock test)
2. [drivers: Missing size_t format specifiers](https://github.com/OlegHahm/RIOT/pull/25)
    I seem to have missed some size_t format specifiers in drivers that still use an unsigned cast in this PR.

**Comment by fzi-haxel** [comment]
> This should hopefully fix the murdock test

Or not... 
I also re-ran native64 tests after the rebase and found even more problems.

I added a new [PR](https://github.com/OlegHahm/RIOT/pull/27) and compiled all examples/tests for avr-rss2, iotlab-m3, sipeed-longan-nano-tft, esp32-wroom-32, native and "native64".

**Comment by fzi-haxel** [comment]
I think this PR has become a bit complicated to review and is probably still a bit too big to be easily discussed. I believe the best way to proceed with the PR is to break it down into smaller PRs that can be merged step by step.
I updated #19890 to reflect the state of this discussion and marked it as WiP.

**Comment by OlegHahm** [comment]
Probably you're right. Should I fix this one?

**Comment by fzi-haxel** [comment]
> Probably you're right. Should I fix this one?

All commits and review change requests in this PR should now be included in other smaller PRs.
The main PRs are
- #20257
- #20271
- #20276

Therefore, I think you could close this PR.
Many thanks to the reviewers and especially to @OlegHahm for the effort.
